Imporove apex

ymd-h opened the pull request:
This PR is for #117 

* Update cpprb and use the new class `MPPrioritizedReplayBuffer` (multi-process version PER), which doesn't requires manual lock
  * `MPPrioritizedReplayBuffer` doesn't lock whole buffer but critical section only
* Remove busy loop and use efficient waiting by queue's `get()` method.
* Remove slow proxy, `SyncManager`
* Reduce access for shared value, `trained_steps`
  * Fix atomic increment, too. (`+=` requires manual lock. See [doc](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Value))
* Run learner at main process instead of background process
* Remove `writer.flush()` (I just let TensorFlow flush. We might need to adjust flush timing for our needs)
* Remove eps (small value) addition to TD, which is added in buffer.

This improvement have larger effect for small network and/or simple `Env`.

I tested by running example/run_apex_dqn.py with default "CartPole-v0" on CPU machine.
Please test other Envs and/or on GPU.


P.S.
Weights distribution with multiple queues seems to be inefficient because of multiple copying.
I will continue to consider other solution.

ymd-h committed:
Use lockless MPPrioritizedReplayBuffer

ymd-h committed:
Remove busy loop but just wait by SimpleQueue.get()

ymd-h committed:
Reduce access to lock object by copying to local object

ymd-h committed:
Pass set of weights and steps to evaluator simultaneously

ymd-h committed:
Remove eps, which added internally

ymd-h committed:
Reduce one process

ymd-h committed:
Require cpprb 9.4.2

ymd-h committed:
F401: Remove unused SyncManager

ymd-h committed:
Fix: E231 missing whitespace after ','

keiohta commented:
Hi @ymd-h , thanks for brilliant PR!! I would really appreciate your continuous support!

I checked changes of codes, and I think all of them would contribute improvement of ApeX performance.

> P.S.
> Weights distribution with multiple queues seems to be inefficient because of multiple copying.
> I will continue to consider other solution.
Yeah, this is true. I also consider some workaround. Thanks!

ymd-h committed:
Merge branch 'master' into Feature_imporove_apex

ymd-h committed:
Fix: __file__ -> __name__ for logger

ymd-h committed:
Use cpprb>=10.5.2

cpprb 10.5.2 imporoves performance of PrioritizedReplayBuffer and
MPPrioritizedReplayBuffer.

ymd-h commented:
@keiohta 

I updated the PR.

* Merge master branch to take recent update
* Use newer cpprb, which improved Segment Tree implementation

On my local machine, (a part of) logs at `example/run_apex_dqn.py` are followings;

Before PR
<img width="600" alt="SS 2022-02-13 11 33 24" src="https://user-images.githubusercontent.com/3405876/153735807-a299e7f8-1960-4739-9bd2-15cb19c196ab.png">

After PR
<img width="600" alt="SS 2022-02-13 11 35 09" src="https://user-images.githubusercontent.com/3405876/153735837-cb20fb30-ceb5-41c0-989b-34427515a73e.png">


It seems that Super Linter v3 is strangely broken and we need to upgrade to v4. (I will add soon.)
https://github.com/github/super-linter/issues/2253

ymd-h committed:
Update Super Linter to version 4

Ref: https://github.com/github/super-linter/issues/2253
